File: benches/fixtures/sample.go
package service

import (
	"context"
	"errors"
	"fmt"
	"net/http"
	"sync"
	"time"
)

const (
	MaxRetries       = 3
	DefaultTimeoutMs = 5000
)

// ServiceError represents a domain-layer error.
type ServiceError struct {
	Message string
	Code    string
	Status  int
}

func (e *ServiceError) Error() string {
	return fmt.Sprintf("[%s] %s", e.Code, e.Message)
}

// Indexable is the interface for items that can be indexed.
type Indexable interface {
	ID() string
	Kind() string
	Score() float64
}

// ClientConfig holds configuration for the HTTP client.
type ClientConfig struct {
	BaseURL         string
	TimeoutMs       int
	MaxRetries      int
	Headers         map[string]string
	UserAgent       string
	FollowRedirects bool
	VerifySSL       bool
	Proxy           string
}

// DefaultClientConfig returns a ClientConfig with sensible defaults.
func DefaultClientConfig(baseURL string) ClientConfig {
	return ClientConfig{
		BaseURL:         baseURL,
		TimeoutMs:       DefaultTimeoutMs,
		MaxRetries:      MaxRetries,
		Headers:         make(map[string]string),
		UserAgent:       "taoki/0.1",
		FollowRedirects: true,
		VerifySSL:       true,
	}
}

// Role represents a user's permission level.
type Role int

const (
	RoleViewer Role = iota
	RoleEditor
	RoleAdmin
)

// User is a domain entity representing an application user.
type User struct {
	ID    string
	Name  string
	Email string
	Role  Role
}

func (u *User) GetID() string    { return u.ID }
func (u *User) Kind() string     { return "user" }
func (u *User) Score() float64   { return 1.0 }

// ClientService provides HTTP-backed service operations.
type ClientService struct {
	mu     sync.Mutex
	config ClientConfig
	client *http.Client
}

// NewClientService creates a new ClientService.
func NewClientService(config ClientConfig) *ClientService {
	return &ClientService{
		config: config,
		client: &http.Client{
			Timeout: time.Duration(config.TimeoutMs) * time.Millisecond,
		},
	}
}

// FetchUser retrieves a user by ID.
func (s *ClientService) FetchUser(ctx context.Context, id string) (*User, error) {
	if id == "" {
		return nil, &ServiceError{Message: "id cannot be empty", Code: "INVALID_INPUT", Status: 400}
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, s.config.BaseURL+"/users/"+id, nil)
	if err != nil {
		return nil, fmt.Errorf("build request: %w", err)
	}
	resp, err := s.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("do request: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode == http.StatusNotFound {
		return nil, &ServiceError{Message: "user not found", Code: "NOT_FOUND", Status: 404}
	}
	return &User{ID: id, Name: "Alice", Email: "alice@example.com"}, nil
}

// PaginatedResult holds a page of results.
type PaginatedResult[T any] struct {
	Items   []T
	Total   int
	Page    int
	PerPage int
}

// Paginate returns a page of items from a slice.
func Paginate[T any](items []T, page, perPage int) PaginatedResult[T] {
	total := len(items)
	start := (page - 1) * perPage
	if start > total {
		start = total
	}
	end := start + perPage
	if end > total {
		end = total
	}
	return PaginatedResult[T]{Items: items[start:end], Total: total, Page: page, PerPage: perPage}
}

// ParseHeader splits a raw "Key: Value" header string.
func ParseHeader(raw string) (string, string, error) {
	for i, c := range raw {
		if c == ':' {
			return trim(raw[:i]), trim(raw[i+1:]), nil
		}
	}
	return "", "", errors.New("invalid header: missing colon")
}

func trim(s string) string {
	start, end := 0, len(s)
	for start < end && (s[start] == ' ' || s[start] == '\t') {
		start++
	}
	for end > start && (s[end-1] == ' ' || s[end-1] == '\t') {
		end--
	}
	return s[start:end]
}

func internalHash(data []byte) uint64 {
	var h uint64 = 0xcbf29ce484222325
	for _, b := range data {
		h ^= uint64(b)
		h *= 0x100000001b3
	}
	return h
}

File: benches/fixtures/sample.js
"use strict";

const EventEmitter = require("events");
const path = require("path");
const axios = require("axios");

const MAX_RETRIES = 3;
const DEFAULT_TIMEOUT_MS = 5000;

let globalInstance = null;

/**
 * @typedef {Object} ClientConfig
 * @property {string} baseUrl
 * @property {number} [timeoutMs]
 * @property {number} [maxRetries]
 * @property {Record<string, string>} [headers]
 */

/**
 * Custom service error.
 */
class ServiceError extends Error {
  constructor(message, code = "INTERNAL", status = 500) {
    super(message);
    this.name = "ServiceError";
    this.code = code;
    this.status = status;
  }
}

/**
 * A simple user model.
 */
class User {
  constructor(id, name, email, role = "viewer") {
    this.id = id;
    this.name = name;
    this.email = email;
    this.role = role;
  }

  toJSON() {
    return { id: this.id, name: this.name, email: this.email, role: this.role };
  }

  static fromJSON(data) {
    return new User(data.id, data.name, data.email, data.role ?? "viewer");
  }
}

/**
 * HTTP client wrapping axios.
 */
class ClientService extends EventEmitter {
  constructor(config) {
    super();
    this.config = config;
    this.client = axios.create({
      baseURL: config.baseUrl,
      timeout: config.timeoutMs ?? DEFAULT_TIMEOUT_MS,
      headers: config.headers ?? {},
    });
  }

  async fetchUser(id) {
    if (!id) throw new ServiceError("id cannot be empty", "INVALID_INPUT", 400);
    const resp = await this.client.get(`/users/${id}`);
    return User.fromJSON(resp.data);
  }

  async createUser(data) {
    const resp = await this.client.post("/users", data);
    return User.fromJSON(resp.data);
  }
}

/**
 * Paginate an array of items.
 * @param {Array} items
 * @param {number} page
 * @param {number} perPage
 */
function paginate(items, page = 1, perPage = 20) {
  const total = items.length;
  const start = (page - 1) * perPage;
  return { items: items.slice(start, start + perPage), total, page, perPage };
}

/**
 * Parse a raw header string into [key, value].
 * @param {string} raw
 */
function parseHeader(raw) {
  const idx = raw.indexOf(":");
  if (idx === -1) return null;
  return [raw.slice(0, idx).trim(), raw.slice(idx + 1).trim()];
}

function _internalHash(data) {
  let h = BigInt("0xcbf29ce484222325");
  for (const b of data) {
    h ^= BigInt(b);
    h = (h * BigInt("0x100000001b3")) & BigInt("0xffffffffffffffff");
  }
  return h;
}

const app = require("express")();

app.use(require("express").json());

app.get("/users/:id", async (req, res) => {
  try {
    const svc = globalInstance;
    const user = await svc.fetchUser(req.params.id);
    res.json(user.toJSON());
  } catch (err) {
    res.status(err.status ?? 500).json({ error: err.message });
  }
});

app.post("/users", async (req, res) => {
  const svc = globalInstance;
  const user = await svc.createUser(req.body);
  res.status(201).json(user.toJSON());
});

globalInstance = new ClientService({ baseUrl: "http://localhost:8080" });

module.exports = { ServiceError, User, ClientService, paginate, parseHeader };
